keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,960 @@
+* At restore time, create a single volume from individual snapshot and then join an existing group.
+ * Create an empty group
+ * Create a volume from snapshot in the group
+
+The following are non-goals:
+* Non goal: Create a new volume group from a list of existing volumes.
+* Non goal: Create a new group and at the same time create a list of new volumes in the group.
+* Non goal: Create a group from an existing group snapshot in one step.
+
+```
+message CreateVolumeGroupRequest {
+  option (alpha_message) = true;
+
+  // suggested name for volume group (required for idempotency)
+  // This field is REQUIRED.
+  string name = 1;
+
+  // params passed from VolumeGroupClass
+  // This field is OPTIONAL.
+  map<string, string> parameters = 2;
+}
+
+message CreateVolumeGroupResponse {
+  option (alpha_message) = true;
+
+  // Contains all attributes of the newly created volume group.
+  // This field is REQUIRED.
+  VolumeGroup volume_group = 1;
+}
+
+message VolumeGroup {
+  option (alpha_message) = true;
+
+  // The identifier for this volume group, generated by the plugin.
+  // This field is REQUIRED.
+  string volume_group_id = 1;
+
+  // Opaque static properties of the volume group.
+  // This field is OPTIONAL.
+  map<string, string> volume_group_context = 2;
+
+  // Underlying volumes in this group. The same definition in CSI Volume.
+  // This field is OPTIONAL.
+  repeated .csi.v1.Volume volumes = 3;
+}
+```
+
+#### CreateVolume
+
+1. When a new volume is created with a volume group id parameter, the volume will be created and added to the existing volume group.
+2. A new volume can also be created without a volume group id parameter. It can be added to a volume group later through the ModifyVolumeGroup RPC.
+
+Note that for filesystems based storage systems, only option 1 can be supported. For block based storage systems. Both option 1 and 2 may be supported. However there is a possibility that option 2 will not work for ConsistencyGroups as the volume is created without the consideration of which group the volume will be placed in.
+
+```
+message CreateVolumeRequest {
+      string name = 1;
+      …
+      repeated string volume_group_id = 8 [(alpha_field) = true];
+}
+```
+
+#### ModifyVolume
+
+This RPC is called when an existing volume is added to an existing volume group.
+A volume group id parameter will be in the ModifyVolumeRequest.
+
+```
+message ModifyVolumeRequest {
+      string volume_id = 1;
+      repeated string volume_group_id = 2 [(alpha_field) = true];
+}
+```
+
+#### DeleteVolumeGroup
+
+```
+message DeleteVolumeGroupRequest {
+  option (alpha_message) = true;
+
+  // The ID of the volume group to be deprovisioned.
+  // This field is REQUIRED.
+  string volume_group_id = 1;
+
+  // Secrets required by plugin to complete volume group deletion request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  map<string, string> secrets = 2 [(csi_secret) = true];
```

Comment:
Interestingly `secrets` are listed in `DeleteVolumeGroup`, but not in `CreateVolumeGroup` or `ModifyVolumeGroup`. Most likely, `secrets` will be needed for most (if not all) operations.

Reply:
That's a good catch.  Secrets should be added to other RPCs.

Reply:
Done.